frontend/src/js/components/PageViewer/VirtualScroll.tsx

```diff
@@ -0,0 +1,82 @@
+import _ from "lodash";
+import React, { FC, ReactNode, useLayoutEffect, useRef, useState } from "react";
+import { CONTAINER_AND_MARGIN_SIZE } from "./model";
+import styles from "./VirtualScroll.module.css";
+
+type VirtualScrollProps = {
+  totalPages: number;
+  initialPage?: number;
+  renderPage: (pageNumber: number) => ReactNode;
+};
+
+export const VirtualScroll: FC<VirtualScrollProps> = ({
+  totalPages,
+  initialPage,
+  renderPage,
+}) => {
+  // Tweaked this and 2 seems to be a good amount on a regular monitor
+  // The fewer pages we preload the faster the initial paint will be
+  // Could possibly make it dynamic based on the visible of the container
+  const PRELOAD_PAGES = 2;
+
+  const pageHeight = CONTAINER_AND_MARGIN_SIZE;
+
+  const viewport = useRef<HTMLDivElement>(null);
+
+  const [topPage, setTopPage] = useState(1);
+  //const [midPage, setMidPage] = useState(1); // Todo hook up to URL
+  const [botPage, setBotPage] = useState(1 + PRELOAD_PAGES);
+
+  const onScroll = () => {
+    if (viewport?.current) {
+      const v = viewport.current;
+
+      const currentMid = v.scrollTop + v.clientHeight / 2;
+
+      const topEdge = currentMid - PRELOAD_PAGES * pageHeight;
```

Comment:
with all of these arithmetic expressions, I'd always prefer explicit brackets even when redundant because of BODMAS - I find it much easier to read brackets than recite BODMAS to myself and then try and mentally insert the brackets